### ISSUE TITLE
t.i.defer.inlineCallbacks documentation issues

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2674]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)| @bitglue reported|
|-|-|
|Trac ID|trac#2674|
|Type|defect|
|Created|2007-05-23 17:16:11Z|
|Branch|https://github.com/twisted/twisted/tree/inlineCallbacks-docs-2674|

inlineCallbacks currently allows one to yield a value that is not a Deferred instance without an error. When this is done, it seems to .send() the yielded value back to the generator immediately. I am not sure this is useful behavior, since if you can yield it, you could assign it. The only one case is maybe functions that might return a deferred sometimes; then "yield" serves a function similar to maybeDeferred.

I don't think that's a very common case. It seems more likely to me that yielding a non-deferred is a mistake and should raise an error. Also, it might be useful to have the capability to yield other things from the generator that instruct inlineCallbacks to do other things. For example, call next() in a thread, or maybe give it to runInteraction of an adbapi.ConnectionPool. This is not possible with the current behavior.

Arguments against the usefulness of the current behavior aside, it should at least be documented, if not depreciated or changed.

Also, I found it somewhat surprising that "returnValue(Deferred())" causes an error:

```
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/2.5/lib/python2.5/site-packages/twisted/internet/defer.py", line 746, in gotResult
    _inlineCallbacks(r, g, deferred)
  File "/Library/Frameworks/Python.framework/Versions/2.5/lib/python2.5/site-packages/twisted/internet/defer.py", line 733, in _inlineCallbacks
    deferred.callback(e.value)
  File "/Library/Frameworks/Python.framework/Versions/2.5/lib/python2.5/site-packages/twisted/internet/defer.py", line 239, in callback
    self._startRunCallbacks(result)
  File "/Library/Frameworks/Python.framework/Versions/2.5/lib/python2.5/site-packages/twisted/internet/defer.py", line 304, in _startRunCallbacks
    self._runCallbacks()
--- <exception caught here> ---
  File "/Library/Frameworks/Python.framework/Versions/2.5/lib/python2.5/site-packages/twisted/internet/defer.py", line 317, in _runCallbacks
    self.result = callback(self.result, *args, **kw)
  File "/Library/Frameworks/Python.framework/Versions/2.5/lib/python2.5/site-packages/twisted/internet/defer.py", line 746, in gotResult
    _inlineCallbacks(r, g, deferred)
  File "/Library/Frameworks/Python.framework/Versions/2.5/lib/python2.5/site-packages/twisted/internet/defer.py", line 733, in _inlineCallbacks
    deferred.callback(e.value)
  File "/Library/Frameworks/Python.framework/Versions/2.5/lib/python2.5/site-packages/twisted/internet/defer.py", line 238, in callback
    assert not isinstance(result, Deferred)
exceptions.AssertionError: 
```

especially since .addCallback(lambda _: Deferred()) is not an error.

Attachments:

* [inlineCallbacks-docs-2674.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bcf/bcf06593eb08b05b503a29a24a12e3acfc7540c8/cf5e3c3608190a2d0047871f5dfdd069acf1d893.patch) (1316 bytes) - added by indigo on 2011-09-30 21:03:38Z - Documentation additions.

<details><summary>Searchable metadata</summary>

```
trac-id__2674 2674
type__defect defect
reporter__indigo indigo
priority__normal normal
milestone__ 
branch__branches_inlineCallbacks_docs_2674 branches/inlineCallbacks-docs-2674
branch_author__indigo indigo
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1179940571000000 1179940571000000
changetime__1317789017000000 1317789017000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph set owner to @bitglue|
|-|-|

"documentation patches welcome"

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)|@bitglue removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

(In [32720]) Branching to 'inlineCallbacks-docs-2674'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/inlineCallbacks-docs-2674)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.  This looks like a good documentation improvement.  I'll leave discussion of whether the functionality should be extended for another ticket. :)

The doc builder is happy, so I'll apply this to trunk.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [32740]) Merge inlineCallbacks-docs-2674

Author: indigo
Reviewer: exarkun
Fixes: [#2674](https://github.com/twisted/twisted/issues/2674)

Expand the `inlineCallbacks` API documentation to explain the behavior of
yielding a non-`Deferred` and that `returnValue` will not accept a `Deferred`.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Isn't it also a bug that this is an `AssertionError` and not something more informative / better documented at runtime?
